src/router_core/address_watch.c

```diff
@@ -127,15 +129,43 @@ static void qdr_core_watch_address_CT(qdr_core_t *core, qdr_action_t *action, bo
         qd_iterator_t *iter = qdr_field_iterator(action->args.io.address);
         qd_iterator_annotate_prefix(iter, action->args.io.address_class);
         qd_iterator_reset_view(iter, ITER_VIEW_ADDRESS_HASH);
+        qdr_address_t *addr = 0;
+
+        qd_hash_retrieve(core->addr_hash, iter, (void**) &addr);
+        if (!addr) {
+            qdr_address_config_t   *addr_config;
+            qd_address_treatment_t  treatment =
+                qdr_treatment_for_address_hash_with_default_CT(core, iter, action->args.io.treatment, &addr_config);
+
+            addr = qdr_address_CT(core, treatment, addr_config);
+            if (!!addr) {
+                qd_hash_insert(core->addr_hash, iter, addr, &addr->hash_handle);
+                DEQ_ITEM_INIT(addr);
+                DEQ_INSERT_TAIL(core->addrs, addr);
+            } else {
+                qd_log(core->log, QD_LOG_CRITICAL, "Failed to create address for watch");
+                assert(false);
+            }
+        }
 
-        qdr_address_watch_t *watch = new_qdr_address_watch_t();
-        ZERO(watch);
-        watch->watch_handle = action->args.io.value32_1;
-        watch->address_hash = (char*) qd_iterator_copy(iter);
-        watch->handler      = action->args.io.watch_handler;
-        watch->context      = action->args.io.context;
-
-        DEQ_INSERT_TAIL(core->addr_watches, watch);
+        if (!!addr) {
+            if (!addr->watch) {
+                qdr_address_watch_t *watch = new_qdr_address_watch_t();
+                ZERO(watch);
+                watch->watch_handle = action->args.io.value32_1;
+                watch->addr         = addr;
+                watch->handler      = action->args.io.watch_handler;
+                watch->context      = action->args.io.context;
+                DEQ_INSERT_TAIL(core->addr_watches, watch);
+
+                addr->watch = watch;
+                addr->ref_count++;
+
+                qdrc_event_addr_raise(core, QDRC_EVENT_ADDR_WATCH_ON, addr);
```

Comment:
Would it be possible to trigger the newly watched address now so the watcher gets the current counters?

Reply:
This sounds like a good idea.

Reply:
I added the requested feature.